sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -1359,17 +1379,27 @@ __ESIMD_API
   return atomic_update<Op, Tx, N>(__ESIMD_DNS::accessorToPointer<Tx>(acc),
                                   offset, mask);
 #else
-  detail::check_atomic<Op, Tx, N, 0>();
-  static_assert(sizeof(Toffset) == 4, "Only 32 bit offset is supported");
+
   if constexpr (Op == atomic_op::load) {
-    return atomic_update<atomic_op::bit_or, Tx, N>(acc, offset, simd<Tx, N>(0),
-                                                   mask);
-  } else {
-    static_assert(sizeof(Tx) == 4, "Only 32 bit data is supported");
-    const auto si = __ESIMD_NS::get_surface_index(acc);
-    using T = typename detail::__raw_t<Tx>;
-    return __esimd_dword_atomic0<Op, T, N>(mask.data(), si, offset.data());
+    if constexpr (std::is_integral_v<Toffset>) {
+      return atomic_update<atomic_op::bit_or, Tx, N>(acc, offset,
+                                                     simd<Tx, N>(0), mask);
+    } else {
+      using Tint = detail::uint_type_t<sizeof(Tx)>;
+      simd<Tint, N> Res = atomic_update<atomic_op::bit_or, Tint, N>(
+          acc, offset, simd<Tint, N>(0), mask);
+      return Res.template bit_cast_view<Tx>();
+    }
   }
+
+  detail::check_atomic<Op, Tx, N, 0>();
+  static_assert(sizeof(Toffset) == 4, "Only 32 bit offset is supported");
+
+  static_assert(sizeof(Tx) == 4, "Only 32 bit data is supported");
+  const auto si = __ESIMD_NS::get_surface_index(acc);
+  using T = typename detail::__raw_t<Tx>;
+  return __esimd_dword_atomic0<Op, T, N>(mask.data(), si, offset.data());
+
```

Comment:
Please put this block to 'else' branch of 'if constexpr (OP == atomic_op::load)' condition.
'if constexpr' works a bit differently than regular 'if', thus this code needs to be in 'else'-branch.